src/JsonRpc/OutputHandler.cs

```diff
@@ -10,80 +16,71 @@ namespace OmniSharp.Extensions.JsonRpc
 {
     public class OutputHandler : IOutputHandler
     {
-        private readonly Stream _output;
+        private readonly PipeWriter _pipeWriter;
         private readonly ISerializer _serializer;
-        private readonly ILogger<OutputHandler> _outputHandler;
-        private readonly Thread _thread;
-        private readonly BlockingCollection<object> _queue;
-        private readonly CancellationTokenSource _cancel;
+        private readonly ILogger<OutputHandler> _logger;
+        private readonly Subject<object> _queue;
         private readonly TaskCompletionSource<object> _outputIsFinished;
+        private readonly CompositeDisposable _disposable;
 
-        public OutputHandler(Stream output, ISerializer serializer, ILogger<OutputHandler> outputHandler)
+        public OutputHandler(PipeWriter pipeWriter, ISerializer serializer, ILogger<OutputHandler> logger)
         {
-            if (!output.CanWrite)
-                throw new ArgumentException($"must provide a writable stream for {nameof(output)}", nameof(output));
-            _output = output;
+            _pipeWriter = pipeWriter;
             _serializer = serializer;
-            _outputHandler = outputHandler;
-            _queue = new BlockingCollection<object>();
-            _cancel = new CancellationTokenSource();
+            _logger = logger;
+            _queue = new Subject<object>();
             _outputIsFinished = new TaskCompletionSource<object>();
-            _thread = new Thread(ProcessOutputQueue) {IsBackground = true, Name = "ProcessOutputQueue"};
+
+            _disposable = new CompositeDisposable {
+                _queue
+                    .Select(value => Observable.FromAsync(ct => ProcessOutputStream(value, ct)))
+                    .Concat()
+                    .ObserveOn(new EventLoopScheduler(_ => new Thread(_) {IsBackground = true, Name = "OutputHandler"}))
+                    .Subscribe(),
+                _queue,
+                Disposable.Create(() => _pipeWriter?.Complete())
+            };
         }
 
-        public void Start()
+        public void Send(object value)
```

Comment:
@TanayParikh this `OutputHandler` class is essentialy what our `AutoFlushingStreamWriter` tries to workaround. Changes  here are a lot like what we'd have to do on our side until we consume these changes 😄